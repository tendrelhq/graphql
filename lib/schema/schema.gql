enum AuthenticationProvider {
  legacy
  clerk
}

enum LanguageCode {
  en
  es
}

type Language {
  id: ID!
  code: LanguageCode!
  name: Name!
  name_id: ID!
}

type Name {
  id: ID!
  language: Language!
  language_id: ID!
  value: String!
}

type Tag {
  id: ID!
  name: Name!
  name_id: ID!
  type: String!
  parent: Tag
  parent_id: ID
}

type Customer {
  id: ID!
  active: Boolean!
  name: Name!
  name_id: ID!
  defaultLanguage: Language!
  default_language_id: ID!
}

type Location {
  id: ID!
  active: Boolean!
  activated_at: String
  deactivated_at: String
  name: Name!
  name_id: ID!
  parent: Location
  parent_id: ID
  scan_code: ID
  site: Location!
  site_id: ID!
  tags: [Tag!]!

  children(options: LocationsQueryOptions): [Location!]!
}

type User {
  id: ID!
  active: Boolean!
  activated_at: String
  deactivated_at: String
  authentication_provider: Tag
  authentication_provider_id: ID
  language: Language!
  language_id: ID!
  name: String!
  tags: [Tag!]!
}

type Worker {
  id: ID!
  active: Boolean!
  activated_at: String
  deactivated_at: String
  language: Language!
  language_id: ID!
  role: Tag!
  role_id: ID!
  scan_code: String
  tags: [Tag!]!
  user: User!
  user_id: ID!
}

type Query {
  customers: [Customer!]!
  customer(id: ID!): Customer!
  languages: [Language!]!
  locations(customerId: ID!, options: LocationsQueryOptions): [Location!]!
  location(id: ID!): Location!
  roles: [Tag!]!
  workers(customerId: ID!, options: WorkersQueryOptions): [Worker!]!
}

input LocationsQueryOptions {
  cornerstone: Boolean
  site: Boolean
}

input WorkersQueryOptions {
  active: Boolean
  site: ID
}

type Mutation {
  # Location CRUD
  createLocation(input: CreateLocationInput!): Location!
  updateLocation(input: UpdateLocationInput!): Location!

  # User
  ## CRUD
  createUser(input: CreateUserInput!): User!
  updateUser(input: UpdateUserInput!): User!
  ## Helpers
  acceptInvite(input: AcceptInviteInput!): User!

  # Worker CRUD
  createWorker(input: CreateWorkerInput!): Worker!
  updateWorker(input: UpdateWorkerInput!): Worker!
}

input CreateLocationInput {
  id: ID
  active: Boolean!
  name: CreateNameInput!
  org_id: ID!
  scan_code: ID
}

input UpdateLocationInput {
  id: ID!
  activated_at: String
  deactivated_at: String
  name: UpdateNameInput
  scan_code: ID
}

input CreateUserInput {
  id: ID
  active: Boolean!
  authentication_provider_id: ID
  authentication_identity: ID
  language_id: ID!
  name: String!
}

input UpdateUserInput {
  id: ID!
  authentication_provider_id: ID
  authentication_identity: ID
  language_id: ID!
  name: String!
}

input CreateWorkerInput {
  id: ID
  active: Boolean!
  language_id: ID!
  org_id: ID!
  role_id: ID!
  scan_code: ID
  user_id: ID!
}

input UpdateWorkerInput {
  id: ID!
  activated_at: String
  deactivated_at: String
  language_id: ID!
  role_id: ID!
  scan_code: ID
}

input CreateNameInput {
  id: ID
  active: Boolean!
  language_id: ID!
  value: String!
}

input UpdateNameInput {
  id: ID!
  activated_at: String
  deactivated_at: String
  language_id: ID!
  value: String!
}

input AcceptInviteInput {
  id: ID!
  authentication_identity: ID!
}