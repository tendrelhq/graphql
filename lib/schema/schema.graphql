enum Language {
  en
  es
}

type Tag {
  id: ID!
  name: String!
  type: String!
  parent: ID
}

type Customer {
  id: ID!
  name: String!
  defaultLanguage: Language!
}

type Location {
  id: ID!
  customerId: ID!
  name: String!
  parentId: ID
  tags: [Tag!]!
}

type Query {
  customers: [Customer!]!
  customer(id: ID!): Customer!
  locations(customerId: ID!, parentId: ID): [Location!]!
  sites(customerId: ID!): [Location!]!
}

type Mutation {
  createCustomer(input: CreateCustomerInput!): ID!
  updateCustomer(input: UpdateCustomerInput!): ID!
  deleteCustomer(id: ID!): ID!
  createLocation(input: CreateLocationInput!): ID!
  updateLocation(input: UpdateLocationInput!): ID!
  deleteLocation(id: ID!): ID!
}

input CreateTagInput {
  id: ID!
  name: String!
  type: String!
  parent: ID
}

input UpdateTagInput {
  id: ID!
  name: String
}

input CreateCustomerInput {
  id: ID!
  name: String!
  defaultLanguage: Language!
}

input UpdateCustomerInput {
  id: ID!
  name: String
  defaultLanguage: Language
}

input CreateLocationInput {
  id: ID!
  customerId: ID!
  name: String!
  tags: [CreateTagInput!]!
}

input UpdateLocationInput {
  id: ID!
  name: String
  tags: [UpdateTagInput!]
}
