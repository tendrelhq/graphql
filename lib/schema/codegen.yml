schema: lib/schema/schema.graphql
generates:
  lib/schema/__generated__/resolvers-types.ts:
    plugins:
      - typescript
      - typescript-resolvers
    config:
      contextType: "@/schema#Context"
      useIndexSignature: true
