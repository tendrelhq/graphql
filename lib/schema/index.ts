export * from "./__generated__/resolvers.generated";
export * from "./__generated__/typeDefs.generated";

export type Context = {
  authScope?: string;
  languageTypeId: number;
};
