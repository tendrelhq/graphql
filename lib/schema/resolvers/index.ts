import type { QueryResolvers } from "@/schema";

export type QueryResolver<T extends keyof QueryResolvers> = NonNullable<
  QueryResolvers[T]
>;
