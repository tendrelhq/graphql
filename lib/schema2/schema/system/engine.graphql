"""
  Components characterize Entities as possessing a particular trait.
  They are just simple structs, holding all data necessary to model that trait.
"""
interface Component {
  id: ID!
}
