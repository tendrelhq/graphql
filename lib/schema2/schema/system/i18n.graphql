scalar Locale

"""
  TODO
"""
type DynamicString {
  locale: Locale!
  value: String!
}
