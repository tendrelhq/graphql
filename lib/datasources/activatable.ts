import type { ID, ResolversTypes } from "@/schema";
import { decodeGlobalId } from "@/schema/system";
import type { WithKey } from "@/util";
import DataLoader from "dataloader";
import type { Request } from "express";
import { match } from "ts-pattern";
import { sql, unionAll } from "./postgres";

export function makeActivatableLoader(_req: Request) {
  return new DataLoader<ID, ResolversTypes["Activatable"] | undefined>(
    async keys => {
      const entities = keys.map(decodeGlobalId);
      const byUnderlyingType = entities.reduce((acc, { type, id }) => {
        if (!acc.has(type)) acc.set(type, []);
        acc.get(type)?.push(id);
        return acc;
      }, new Map<string, string[]>());

      const qs = [...byUnderlyingType.entries()].flatMap(([type, ids]) =>
        match(type)
          .with(
            "workresult",
            () => sql`
                (
                  WITH cte AS (
                      SELECT
                          wr.id AS _key,
                          encode(('workresult:' || wr.id || ':active')::bytea, 'base64') AS id,
                          (
                              wr.workresultenddate IS NULL
                              OR wr.workresultenddate > now()
                          ) AS active,
                          wr.workresultstartdate,
                          wr.workresultenddate
                      FROM public.workresult AS wr
                      WHERE wr.id IN ${sql(ids)}
                  )

                  SELECT
                      _key,
                      id,
                      active,
                      (
                          SELECT row_to_json(t)
                          FROM (
                              'Instant' AS "__typename",
                              (extract(epoch from workresultstartdate) * 1000)::text AS "epochMilliseconds"
                          ) t
                      )
                  FROM cte
                  WHERE active = true
                  UNION ALL
                  SELECT
                      _key,
                      id,
                      active,
                      (
                          SELECT row_to_json(t)
                          FROM (
                              'Instant' AS "__typename",
                              (extract(epoch from workresultenddate) * 1000)::text AS "epochMilliseconds"
                          ) t
                      )
                  FROM cte
                  WHERE active = false
                )
            `,
          )
          .with(
            "worktemplate",
            () => sql`
                (
                  WITH cte AS (
                      SELECT
                          id AS _key,
                          encode(('worktemplate:' || id || ':active')::bytea, 'base64') AS id,
                          (
                              worktemplateenddate IS NULL
                              OR worktemplateenddate > now()
                          ) AS active,
                          worktemplatestartdate,
                          worktemplateenddate
                      FROM public.worktemplate
                      WHERE id IN ${sql(ids)}
                  )

                  SELECT
                      _key,
                      id,
                      active,
                      (
                          SELECT row_to_json(t)
                          FROM (
                              SELECT
                                  'Instant' AS "__typename",
                                  (extract(epoch from worktemplatestartdate) * 1000)::text AS "epochMilliseconds"
                          ) t
                      ) AS "updatedAt"
                  FROM cte
                  WHERE active = true
                  UNION ALL
                  SELECT
                      _key,
                      id,
                      active,
                      (
                          SELECT row_to_json(t)
                          FROM (
                              SELECT
                                  'Instant' AS "__typename",
                                  (extract(epoch from worktemplateenddate) * 1000)::text AS "epochMilliseconds"
                          ) t
                      ) as "updatedAt"
                  FROM cte
                  WHERE active = false
                )
            `,
          )
          .otherwise(() => []),
      );

      if (!qs.length) return entities.map(() => undefined);

      type X = WithKey<ResolversTypes["Activatable"]>;
      const xs = await sql<X[]>`${unionAll(qs)}`;
      return entities.map(e => xs.find(x => e.id === x._key));
    },
  );
}
