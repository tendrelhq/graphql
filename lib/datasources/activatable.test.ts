import { describe, expect, test } from "bun:test";
import { encodeGlobalId } from "@/schema/system";
import { makeActiveLoader } from "./activatable";

// biome-ignore lint/suspicious/noExplicitAny:
const makeLoader = () => makeActiveLoader({} as any);

describe.skipIf(!!process.env.CI)("active loader", () => {
  test("load", async () => {
    const data = await makeLoader().load(
      "d29ya3RlbXBsYXRlOjgyN2RlNDE5LTQ0YmMtNGExMi1iOWFmLWI2NDFmYjU3NjQyOQ==",
    );
    expect(data).toMatchSnapshot();
  });

  test("no underlying types", async () => {
    const data = await makeLoader().load(
      // workinstances aren't activatable
      encodeGlobalId({ type: "workinstance", id: "foobar" }),
    );

    expect(data).toBeUndefined();
  });
});
