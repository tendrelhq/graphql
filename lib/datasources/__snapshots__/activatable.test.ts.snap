// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`status loader activate 1`] = `
{
  "_key": "827de419-44bc-4a12-b9af-b641fb576429",
  "active": true,
  "id": "d29ya3RlbXBsYXRlOjgyN2RlNDE5LTQ0YmMtNGExMi1iOWFmLWI2NDFmYjU3NjQyOTphY3RpdmU=",
  "updatedAt": {
    "__typename": "Instant",
    "epochMilliseconds": "1727121569246.000000",
  },
}
`;

exports[`status loader deactivate 1`] = `
{
  "_key": "827de419-44bc-4a12-b9af-b641fb576429",
  "active": true,
  "id": "d29ya3RlbXBsYXRlOjgyN2RlNDE5LTQ0YmMtNGExMi1iOWFmLWI2NDFmYjU3NjQyOTphY3RpdmU=",
  "updatedAt": {
    "__typename": "Instant",
    "epochMilliseconds": "1727121569246.000000",
  },
}
`;

exports[`activatable loader activate 1`] = `
{
  "_key": "827de419-44bc-4a12-b9af-b641fb576429",
  "active": true,
  "id": "d29ya3RlbXBsYXRlOjgyN2RlNDE5LTQ0YmMtNGExMi1iOWFmLWI2NDFmYjU3NjQyOTphY3RpdmU=",
  "updatedAt": {
    "__typename": "Instant",
    "epochMilliseconds": "1727121569246.000000",
  },
}
`;

exports[`activatable loader deactivate 1`] = `
{
  "_key": "827de419-44bc-4a12-b9af-b641fb576429",
  "active": true,
  "id": "d29ya3RlbXBsYXRlOjgyN2RlNDE5LTQ0YmMtNGExMi1iOWFmLWI2NDFmYjU3NjQyOTphY3RpdmU=",
  "updatedAt": {
    "__typename": "Instant",
    "epochMilliseconds": "1727121569246.000000",
  },
}
`;
